eclair-core/src/main/scala/fr/acinq/eclair/wire/Onion.scala

```diff
@@ -17,49 +17,133 @@
 package fr.acinq.eclair.wire
 
 import fr.acinq.bitcoin.ByteVector32
-import fr.acinq.eclair.ShortChannelId
 import fr.acinq.eclair.crypto.Sphinx
-import scodec.bits.{BitVector, ByteVector}
+import fr.acinq.eclair.wire.CommonCodecs._
+import fr.acinq.eclair.wire.OnionTlv._
+import fr.acinq.eclair.wire.TlvCodecs._
+import fr.acinq.eclair.{ShortChannelId, UInt64}
+import scodec.bits.{BitVector, ByteVector, HexStringSyntax}
 import scodec.codecs._
 import scodec.{Codec, DecodeResult, Decoder}
 
 /**
   * Created by t-bast on 05/07/2019.
   */
 
+/**
+  * Tlv types used inside onion messages.
+  */
+sealed trait OnionTlv extends Tlv
+
 case class OnionRoutingPacket(version: Int,
                               publicKey: ByteVector,
                               payload: ByteVector,
                               hmac: ByteVector32)
 
-case class PerHopPayload(shortChannelId: ShortChannelId,
-                         amtToForward: Long,
-                         outgoingCltvValue: Long)
+case class OnionForwardInfo(shortChannelId: ShortChannelId,
+                            amtToForward: Long,
+                            outgoingCltvValue: Long)
+
+case class OnionPaymentInfo(amountMsat: Long, cltvExpiry: Long)
+
+case class OnionPerHopPayload(payload: Either[TlvStream[OnionTlv], OnionForwardInfo]) {
+
+  lazy val paymentInfo: Option[OnionPaymentInfo] = payload match {
+    case Right(OnionForwardInfo(_, amount, cltv)) => Some(OnionPaymentInfo(amount, cltv))
+    case Left(tlv) =>
+      tlv.records.foldLeft((None, None): (Option[Long], Option[Long])) {
+        case ((_, cltv), AmountToForward(amount)) => (Some(amount), cltv)
+        case ((amount, _), OutgoingCltv(cltv)) => (amount, Some(cltv))
+        case (res, _) => res
+      } match {
+        case (Some(amount), Some(cltv)) => Some(OnionPaymentInfo(amount, cltv))
+        case _ => None
+      }
+  }
```

Comment:
Something like:
```scala
      for {
        amount <- tlv.records.collectFirst { case AmountToForward(amountMsat) => amountMsat }
        cltv <- tlv.records.collectFirst { case OutgoingCltv(cltv) => cltv }
      } yield OnionPaymentInfo(amount, cltv)
```
would be easier to read, we go through the records twice but it should not be a performance concern